EIPS/eip-5058.md

```diff
@@ -0,0 +1,205 @@
+
+/**
+ * @dev ERC-721 Non-Fungible Token Standard, optional lockable extension
+ * ERC721 Token that can be locked for a certain period and cannot be transferred.
+ * This is designed for a non-escrow staking contract that comes later to lock a user's NFT
+ * while still letting them keep it in their wallet.
+ * This extension can ensure the security of user tokens during the staking period.
+ * If the nft lending protocol is compatible with this extension, the trouble caused by the NFT
+ * airdrop can be avoided, because the airdrop is still in the user's wallet
+ */
+interface IERC721Lockable is IERC721 {
+    /**
+     * @dev Emitted when `tokenId` token is locked by `operator` from `from`.
+     */
+    event Locked(address indexed operator, address indexed from, uint256 indexed tokenId, uint256 expired);
+
+    /**
+     * @dev Emitted when `tokenId` token is unlocked by `operator` from `from`.
+     */
+    event Unlocked(address indexed operator, address indexed from, uint256 indexed tokenId);
+
+    /**
+     * @dev Emitted when `owner` enables `approved` to lock the `tokenId` token.
+     */
+    event LockApproval(address indexed owner, address indexed approved, uint256 indexed tokenId);
+
+    /**
+     * @dev Emitted when `owner` enables or disables (`approved`) `operator` to lock all of its tokens.
+     */
+    event LockApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+    /**
+     * @dev Returns the locker who is locking the `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - `tokenId` must exist.
+     */
+    function lockerOf(uint256 tokenId) external view returns (address locker);
+
+    /**
+     * @dev Lock `tokenId` token until the block number is greater than `expired` to be unlocked.
+     *
+     * Requirements:
+     *
+     * - `from` cannot be the zero address.
+     * - `tokenId` token must be owned by `from`.
+     * - `expired` must be greater than block.number
+     * - If the caller is not `from`, it must be approved to lock this token
+     * by either {lockApprove} or {setLockApprovalForAll}.
+     *
+     * Emits a {Locked} event.
+     */
+    function lockFrom(
+        address from,
+        uint256 tokenId,
+        uint256 expired
+    ) external;
+
+    /**
+     * @dev Unlock `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - `from` cannot be the zero address.
+     * - `tokenId` token must be owned by `from`.
+     * - the caller must be the operator who locks the token by {lockFrom}
+     *
+     * Emits a {Unlocked} event.
+     */
+    function unlockFrom(address from, uint256 tokenId) external;
+
+    /**
+     * @dev Gives permission to `to` to lock `tokenId` token.
+     *
+     * Requirements:
+     *
+     * - The caller must own the token or be an approved lock operator.
+     * - `tokenId` must exist.
+     *
+     * Emits an {LockApproval} event.
+     */
+    function lockApprove(address to, uint256 tokenId) external;
```

Comment:
Don't we need to have a parameter expired for this function?

Reply:
It has been considered that, similar to the allowance of erc20, different applications can lock some expired time, which is actually not very common. Therefore, the fully authorized method of expired is adopted, and the locked time is given to a trusted third party.